fl4health/server/base_server.py

```diff
@@ -32,42 +30,56 @@ def __init__(
+            client_manager (ClientManager): Determines the mechanism by which clients
+                are sampled by the server, if they are to be sampled at all.
+            strategy (Optional[Strategy], optional): The aggregation strategy to be
+                used by the server to handle. client updates and other information
+                potentially sent by the participating clients. If None the strategy is
+                FedAvg as set by the flwr Server.
+            reporters (Sequence[BaseReporter], optional): A sequence of FL4Health
+                reporters which the server should send data to before and after each round.
+            checkpointer (TorchCheckpointer | Sequence [TorchCheckpointer], optional):
+                To be provided if the server should perform server side checkpointing
+                based on some criteria. If none, then no server-side checkpointing is
+                performed. Multiple checkpointers can also be passed in a sequence to
+                checkpointer based on multiple criteria. Ensure checkpoint names are
+                different for each checkpoint or they will overwrite on another.
                 Defaults to None.
-            checkpointer (Optional[Union[TorchCheckpointer, Sequence [TorchCheckpointer]]], optional): To be provided
-                if the server should perform server side checkpointing based on some criteria. If none, then no
-                server-side checkpointing is performed. Multiple checkpointers can also be passed in a sequence to
-                checkpointer based on multiple criteria. Ensure checkpoint names are different for each checkpoint
-                or they will overwrite on another. Defaults to None.
-            metrics_reporter (Optional[MetricsReporter], optional): A metrics reporter instance to record the metrics
-                during the execution. Defaults to an instance of MetricsReporter with default init parameters.
-            server_name (Optional[str]): An optional string name to uniquely identify server.
+            server_name (Optional[str]): An optional string name to uniquely identify
+                server.
         """
 
         super().__init__(client_manager=client_manager, strategy=strategy)
-        self.wandb_reporter = wandb_reporter
         self.checkpointer = [checkpointer] if isinstance(checkpointer, TorchCheckpointer) else checkpointer
         self.server_name = server_name if server_name is not None else generate_hash()
 
-        if metrics_reporter is not None:
-            self.metrics_reporter = metrics_reporter
-        else:
-            self.metrics_reporter = MetricsReporter()
+        self.reporters = [] if reporters is None else list(reporters)
+        for r in self.reporters:
+            r.initialize(id=self.server_name)
+
+    def report_centralized_eval(self, history: History, num_rounds: int) -> None:
```

Comment:
This isn't really state tracking so much as it is parsing the server specific history object. Since the centralised metrics can not be returned by fit or eval methods (they are the outputs of an eval function passed to the strategy) they can only be obtained by either parsing this history object, calling the strategy.eval function manually (which would cause it to run twice) or overriding the flwr.server.fit parent method entirely. I opted to parse the history object. The parsing would be a lot simpler if they hadn't packages the metrics as list of 2d tuples with the metric value itself and the server round.

Reply:
given that the server runs the eval every round, packaging the round explicitly seems unnecessary, but it is what it is

Reply:
Anyways since this functionality is specific to the server class, i don't think it can be moved into the report manager. Resolving for now, but open to discussion if you have any suggestions

Reply:
Yeah I think this is a good argument against it. Sounds good.